it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,244 @@
+import com.google.cloud.spanner.InstanceInfo;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Spanner;
+import com.google.cloud.spanner.SpannerOptions;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default class for implementation of {@link SpannerResourceManager} interface.
+ *
+ * <p>The class supports one instance, one database, and multiple tables per manager object. The
+ * instance and database are created when the manager object is.
+ *
+ * <p>The instance and database ids are formed using testId. The database id will be {testId}, with
+ * some extra formatting. The instance id will be "{testId}-{ISO8601 time, microsecond precision}",
+ * with additional formatting. Note: If testId is more than 30 characters, a new testId will be
+ * formed for naming: {first 21 chars of long testId} + “-” + {8 char hash of testId}.
+ *
+ * <p>The class is thread-safe.
+ */
+public class DefaultSpannerResourceManager implements SpannerResourceManager {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultSpannerResourceManager.class);
+  private static final int MAX_BASE_ID_LENGTH = 30;
+
+  private boolean hasInstance = false;
+
+  private final String projectId;
+  private final String instanceId;
+  private final String databaseId;
+  private final String region;
+
+  private final Spanner spanner;
+  private final InstanceAdminClient instanceAdminClient;
+  private final DatabaseAdminClient databaseAdminClient;
+  private DatabaseClient databaseClient;
+
+  private DefaultSpannerResourceManager(Builder builder) {
+    this(
+        SpannerOptions.newBuilder().setProjectId(builder.projectId).build().getService(),
+        builder.testId,
+        builder.projectId,
+        builder.region);
+  }
+
+  @VisibleForTesting
+  DefaultSpannerResourceManager(Spanner spanner, String testId, String projectId, String region) {
+    if (testId.length() > MAX_BASE_ID_LENGTH) {
+      testId = generateNewId(testId, MAX_BASE_ID_LENGTH);
+    }
+    this.projectId = projectId;
+    this.instanceId = generateInstanceId(testId);
+    this.databaseId = generateDatabaseId(testId);
+
+    this.region = region;
+    this.spanner = spanner;
+    this.instanceAdminClient = spanner.getInstanceAdminClient();
+    this.databaseAdminClient = spanner.getDatabaseAdminClient();
+  }
+
+  public static Builder builder(String testId, String projectId, String region) {
+    checkArgument(testId.length() >= 2, "testId length must be at least 2!");
+    checkArgument(Character.isLetter(testId.charAt(0)), "testId must start with a letter!");
```

Comment:
@zhoufek 

Should I static import isLetter() here and isDigit() in the Utils class since the names are self-explanatory?

Reply:
Okay